### PR TITLE
Add more options to the Analysis widget

### LINKED ISSUE #875: Add anal.in in Advanced Analysis
**Is your feature request related to a problem? Please describe.**
To select a particular zone to analyze

**Describe the solution you'd like**
Add anal.in selector in advanced analysis (similar to #395, #873)

### PR DESCRIPTION
**Pull request by ocastejon**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [x] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [x] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)

**Detailed description**

This MR adds the remaining options to the Analysis Widget that were proposed by @ITAYC0HEN in https://hackmd.io/@Megabeets/H1jBD3Hxv/edit, which are:

- `anal.in`
- `anal.prelude`
- `anal.ptrdepth`

For the `anal.in` option, I selected the following options only:

- All executable maps (`io.maps.x`)
- All maps (`io.maps`)
- Current map (`io.map`)
- Raw (`raw`)
- Current block (`block`)
- Current mapped section (`bin.section`)
- All mapped sections (`bin.sections`)

However, if you think other options should be there, I can add them too.

**Test plan (required)**

1. Open Cutter and go to Edit -> Preferences - > Analysis
2. Select one or more options in the drop-down "Search boundaries for analysis" and check with the console that the value of `anal.in` changes accordingly
3. Change the value of the option "Pointer depth" and check with the console that the value of `anal.ptrdepth` changes accordingly
4. Change the value of the option "Prelude" and check with the console that the value of `anal.prelude` changes accordingly
5. Close and reopen the dialog, and check that the last selected values for these options are maintained

anal-options-2

**Closing issues**

Closes #875

### CODE REVIEW

File: src/dialogs/preferences/AnalOptionsWidget.ui [hunk 1]
```diff
@@ -133,6 +133,125 @@
...
+          <item row="0" column="1">
+           <widget class="QLineEdit" name="preludeLineEdit">
+            <property name="sizePolicy">
+             <sizepolicy hsizetype="Preferred" vsizetype="Fixed">
+              <horstretch>0</horstretch>
+              <verstretch>0</verstretch>
+             </sizepolicy>
+            </property>
+            <property name="placeholderText">
+             <string notr="true">e.g.: 0x554889e5</string>
```
Comment (ITAYC0HEN) [inline src/dialogs/preferences/AnalOptionsWidget.ui:249]: Are you sure it is `0xa1b2c3` and not `a1 b2 c3`? I have no idea personally
```suggestion
             <string notr="true">e.g.: 0x554889e5</string>
```

Comment (ITAYC0HEN) [reply src/dialogs/preferences/AnalOptionsWidget.ui:249]: Also, I think it is very wise to add the place holder here, good job!

**Review by ITAYC0HEN — CHANGES_REQUESTED** [review]
Very nice job! :))
I suggested some changes

File: src/dialogs/preferences/AnalOptionsWidget.h [hunk 2]
```diff
@@ -41,6 +46,24 @@ private slots:
...
+     * @brief A slot to update the value of anal.ptrdepth when a new value is selected
+     * @param value - The new value for anal.ptrdepth
+     */
+    static void on_ptrdepthSpinBox_valueChanged(int value);
+
+    /**
+     * @brief slot to update the value of anal.prelude when a new value is introduced in the corresponding textbox
+     * @param prelude - The new value for anal.prelude
+     */
+    static void on_preludeLineEdit_textChanged(const QString &prelude);
```
[the remark this answers is not in the document]
Comment (ocastejon) [reply src/dialogs/preferences/AnalOptionsWidget.h]: You're totally right, I missed it. To fix it, in some cases I had to cast the signal to the right type (using `static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged)` for instance), which seems quite ugly to me. I saw that this was being done also in other parts of the code, so I guessed it should be all right.

File: src/dialogs/preferences/AnalOptionsWidget.ui [hunk 3]
```diff
@@ -133,6 +133,125 @@
...
+              <height>0</height>
+             </size>
+            </property>
+            <property name="text">
+             <string>Pointer depth (anal.ptrdepth):</string>
+            </property>
+           </widget>
+          </item>
+          <item row="0" column="1">
+           <widget class="QSpinBox" name="ptrdepthSpinBox">
```
[the remark this answers is not in the document]
Comment (ocastejon) [reply src/dialogs/preferences/AnalOptionsWidget.ui]: Yep, I think so. Changed!

File: src/dialogs/preferences/AnalOptionsWidget.ui (hunk 1 above)
Comment (ocastejon) [reply src/dialogs/preferences/AnalOptionsWidget.ui:249]: To be honest, I also had no idea, but according to the r2book this is the way:
``e anal.prelude = 0x554889e5``
(see the end of the section [Recursive Analysis](https://radare.gitbooks.io/radare2book/content/analysis/code_analysis.html#recursive-analysis))

File: docs/source/user-docs/preferences/analysis.rst [hunk 4]
```diff
@@ -72,3 +86,15 @@ Print warnings encountered while preforming type analysis. These warnings can he
...
 
 **Configuration variable:** ``anal.types.verbose``
+
+Pointer depth
+~~~~~~~~~~~~~
+The maximum number of nested pointers to follow in analysis.
+
+**Configuration variable:** ``anal.ptrdepth``
+
+Prelude
```
[the remark this answers is not in the document]
Comment (ocastejon) [reply docs/source/user-docs/preferences/analysis.rst]: I also updated the label in the UI:

image

File: src/dialogs/preferences/AnalOptionsWidget.ui (hunk 1 above)
Comment (ITAYC0HEN) [reply src/dialogs/preferences/AnalOptionsWidget.ui:249]: good enough :)

File: src/dialogs/preferences/AnalOptionsWidget.cpp [hunk 5]
```diff
@@ -33,7 +45,13 @@ AnalOptionsWidget::AnalOptionsWidget(PreferencesDialog *dialog)
 
     ui->analyzePushButton->setToolTip("Analyze the program using radare2's \"aaa\" command");
     auto *mainWindow = new MainWindow(this);
-    connect(ui->analyzePushButton, &QPushButton::clicked, mainWindow, &MainWindow::on_actionAnalyze_triggered);
+    connect(ui->analyzePushButton, &QPushButton::clicked, mainWindow,
+            &MainWindow::on_actionAnalyze_triggered);
+    connect(ui->analInComboBox, static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
+            this, &AnalOptionsWidget::updateAnalIn);
+    connect(ui->ptrDepthSpinBox, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged), this,
```
Comment (karliss) [inline src/dialogs/preferences/AnalOptionsWidget.cpp:52]: For overloaded signals you can do this, which is slightly less verbose.
```suggestion
    connect<void (QSpinBox::*)(int)>(ui->ptrDepthSpinBox, &QSpinBox::valueChanged, this,
```

Comment (ocastejon) [reply src/dialogs/preferences/AnalOptionsWidget.cpp]: Thanks, that is indeed somewhat nicer to read. I changed both the `ptrDepthSpinBox` and the `analInComboBox`. Thanks!

**Review by ITAYC0HEN — APPROVED** [review]

**Comment by ITAYC0HEN** [comment]
Good work Oriol!

**Pull request merged**